server/src/main/java/org/elasticsearch/common/util/SetBackedScalingCuckooFilter.java

```diff
@@ -0,0 +1,361 @@
+ * These filters are approximate similar to Bloom Filters.
+ *
+ * This datastructure scales as more values are inserted by growing the list of CuckooFilters.
+ * Final size is dependent on the cardinality of data inserted, and the precision specified.
+ */
+public class SetBackedScalingCuckooFilter implements Writeable {
+
+    private static final int FILTER_CAPACITY = 1000000;
+
+    // Package-private for testing
+    Set<MurmurHash3.Hash128> hashes;
+    List<CuckooFilter> filters;
+
+    private final int threshold;
+    private final Random rng;
+    private final int capacity;
+    private final double fpp;
+    private Consumer<Long> breaker = aLong -> {
+        //noop
+    };
+    private boolean isSetMode = true;
+
+    /**
+     * @param threshold The number of distinct values that should be tracked
+     *                  before converting to an approximate representation
+     * @param rng A random number generator needed for the cuckoo hashing process
+     * @param fpp the false-positive rate that should be used for the cuckoo filters.
+     */
+    public SetBackedScalingCuckooFilter(int threshold, Random rng, double fpp) {
+        this.hashes = new HashSet<>(threshold);
+        this.threshold = threshold;
+        this.rng = rng;
+        this.capacity = FILTER_CAPACITY;
+        this.fpp = fpp;
+    }
+
+    public SetBackedScalingCuckooFilter(StreamInput in, Random rng) throws IOException {
+        this.threshold = in.readVInt();
+        this.isSetMode = in.readBoolean();
+        this.rng = rng;
+        this.capacity = in.readVInt();
+        this.fpp = in.readDouble();
+
+        if (isSetMode) {
+            this.hashes = in.readSet(in1 -> {
+                MurmurHash3.Hash128 hash = new MurmurHash3.Hash128();
+                hash.h1 = in1.readZLong();
+                hash.h2 = in1.readZLong();
+                return hash;
+            });
+        } else {
+            this.filters = in.readList(in12 -> new CuckooFilter(in12, rng));
+        }
+    }
+
+    public SetBackedScalingCuckooFilter(SetBackedScalingCuckooFilter other) {
+        this.threshold = other.threshold;
+        this.isSetMode = other.isSetMode;
+        this.rng = other.rng;
+        this.breaker = other.breaker;
+        this.capacity = other.capacity;
+        this.fpp = other.fpp;
+        if (isSetMode) {
+            this.hashes = new HashSet<>(other.hashes);
+        } else {
+            this.filters = new ArrayList<>(other.filters);
+        }
+    }
+
+    /**
+     * Registers a circuit breaker with the datastructure.
+     *
+     * CuckooFilter's can "saturate" and refuse to accept any new values.  When this happens,
+     * the datastructure scales by adding a new filter.  This new filter's bytes will be tracked
+     * in the registered breaker when configured.
+     */
+    public void registerBreaker(Consumer<Long> breaker) {
+        this.breaker = breaker;
+        breaker.accept(getSizeInBytes());
```

Comment:
The breaker defaults to a no-op Consumer: https://github.com/elastic/elasticsearch/pull/35718/files/d9998d59d6c368f9ed4f57a099aa091f1ebac147#diff-fcea5bc0639003e5f340598671d7a8e7R65, but that does assume no one accidentally uses the setter with `null`.  I'll add a null check there.

The reason for this strange arrangement is because we don't always have a breaker available.  E.g. the breaker is available in the aggregators, but when the result is serialized over the wire, we no longer have a breaker because were "outside" the normal aggregator life cycle when doing the reductions.

I might be missing something, but I didn't see a way to grab the breaker service after serialization to the coordinating node, in the Internal* result classes.